### PR TITLE
solver for multi-asset CFMM

### PR DESCRIPTION
**Pull request by AlpinYukseloglu**
Closes: #1369

## What is the purpose of the change

This PR implements the solver for our multi-asset CFMM (as outlined in the previous PR for issue #1369).

This is an expansion of the original two-asset CFMM we extended from Solidly, which we expand here to accommodate n-asset pools without significantly increasing complexity or runtime. While this type of multi-asset pool was one that we found was perhaps unnecessarily complex for volatile assets, it is critical from a capital efficiency standpoint that we have the option for Curve-like 3pools and 4pools to not fragment stablecoin liquidity (and therefore also LP incentives) within our stableswap, especially as we compete for stablecoin liquidity with many other stableswaps.

Our final multi-asset CFMM is `xyz(x^2 + y^2 + w) = k`, where z represents the product of the reserves of all assets that are not being swapped between, and w represents the sum of all of their squares. This PR implements a solver that finds the correct output for of a requested token `a` given an input of another token `b` in the pool and accommodates swaps between any two assets `x` and `y` in an n-asset stableswap. The polynomial we solve for can be described as:

`(x - a)(y + b)z((x - a)^2 + (y +b)^2 + w)`, where we solve for `a`. The solution is quite complex and not easy to visually parse without the abstractions we have laid out in the code comments, so please refer to those for more comprehensive and readable implementation details.

## Brief change log

  - Implemented `solveCfmmMulti` function in `x/gamm/pool-models/stableswap/amm.go` using the derivations set up in code comments
  - Implemented binary search logic for this new multi-asset CFMM
  - Set up tests for multi-asset pools and expanded existing two-asset stableswap tests to also check consistency between our two CFMM implementations
  - Cleaned up/simplified the math for the solver to make it more readable and simpler to compute (e.g. limiting the max bitlen to avoid risking integer overflows)

## Testing and Verifying

The tests for this change line up in coverage with those for our two-asset stableswap CFMM and can be found in `amm_test.go`.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (no)
  - How is the feature or change documented? (not documented)

### CODE REVIEW

File: x/gamm/pool-models/stableswap/amm_test.go [hunk 1]
```diff
@@ -42,10 +49,18 @@ func TestCFMMInvariant(t *testing.T) {
...
 		fmt.Println(xOut)
 		k1 := cfmmConstant(test.xReserve.Sub(xOut), test.yReserve.Add(test.yIn))
 		decApproxEq(t, k0, k1, kErrTolerance)
+
+		// multi-asset stableswap tests
+		k2 := cfmmConstantMulti(test.xReserve, test.yReserve, test.uReserve, test.wSumSquares)
+		xOut2 := solveCfmmMulti(test.xReserve, test.yReserve, test.wSumSquares, test.yIn)
+		fmt.Println(xOut2)
+		k3 := cfmmConstantMulti(test.xReserve.Sub(xOut2), test.yReserve.Add(test.yIn), test.uReserve, test.wSumSquares)
+		decApproxEq(t, k2, k3, kErrTolerance)
```
Comment (AlpinYukseloglu) [inline x/gamm/pool-models/stableswap/amm_test.go:64]: @ValarDragon this is the specific test that is still failing (as discussed on slack)

File: x/gamm/pool-models/stableswap/amm.go [hunk 2]
```diff
@@ -147,6 +149,112 @@ func solveCfmm(xReserve, yReserve, yIn sdk.Dec) sdk.Dec {
...
+	// 					4*(3 (b + y)^2 (b^2 + 2 b y + y^2 + w))^3)
+	// 		 )^(1/3)
+	// Factor (b+y)^2 out from within the square root:
+	// foo = (S2 + (b + y)^2
+	// 			* sqrt((-27 x y (x^2 + y^2 + w))^2 +
+	// 					4 * ((b + y)^2) * (3 (b^2 + 2 b y + y^2 + w))^3)
+	// 		 )^(1/3)
+	sqrt_inner_term1 := x.Mul(y).MulInt64(-27).Mul(s2_inner)                                 // (-27 x y (x^2 + y^2 + w))
+	sqrt_inner_term1_2 := sqrt_inner_term1.Mul(sqrt_inner_term1)                             // sqrt_inner_term1^2
+	s1_inner_3 := (s1_inner.MulInt64(3)).Mul(s1_inner.MulInt64(3)).Mul(s1_inner.MulInt64(3)) // (3 (b^2 + 2 b y + y^2 + w))^3
```
Comment (ValarDragon) [inline x/gamm/pool-models/stableswap/amm.go:231]: Lets compute as `27 s1_inner^3`?

Comment (AlpinYukseloglu) [reply x/gamm/pool-models/stableswap/amm.go]: Good point - added in recent commit!

File: x/gamm/pool-models/stableswap/amm.go [hunk 3]
```diff
@@ -180,6 +288,29 @@ func solveCFMMBinarySearch(constantFunction func(sdk.Dec, sdk.Dec) sdk.Dec) func
...
+				x_high_est = x_est
+			} else if cur_k.LT(k) {
+				x_low_est = x_est
+			}
+			x_est = (x_high_est.Add(x_low_est)).Quo(twodec)
+			cur_k = constantFunction(x_est, yf, uReserve, wSumSquares)
+		}
+		return xReserve.Sub(x_est)
+	}
+}
```
Comment (AlpinYukseloglu) [inline x/gamm/pool-models/stableswap/amm.go:312]: @ValarDragon Could you double-check this implementation?

File: x/gamm/pool-models/stableswap/amm_test.go [hunk 4]
```diff
@@ -16,22 +16,28 @@ func decApproxEq(t *testing.T, exp sdk.Dec, actual sdk.Dec, errTolerance sdk.Dec
...
 	kErrTolerance := sdk.OneDec()
 
 	tests := []struct {
-		xReserve sdk.Dec
-		yReserve sdk.Dec
-		yIn      sdk.Dec
+		xReserve    sdk.Dec
+		yReserve    sdk.Dec
+		uReserve    sdk.Dec
+		wSumSquares sdk.Dec
```
Comment (ValarDragon) [inline x/gamm/pool-models/stableswap/amm_test.go:26]: Lets remove these args, and just hard code the 0 and 1

File: x/gamm/pool-models/stableswap/amm_test.go [hunk 5]
```diff
@@ -42,10 +48,60 @@ func TestCFMMInvariant(t *testing.T) {
...
 
 	for _, test := range tests {
+		// using two-asset cfmm
 		k0 := cfmmConstant(test.xReserve, test.yReserve)
 		xOut := solveCfmm(test.xReserve, test.yReserve, test.yIn)
 		fmt.Println(xOut)
 		k1 := cfmmConstant(test.xReserve.Sub(xOut), test.yReserve.Add(test.yIn))
 		decApproxEq(t, k0, k1, kErrTolerance)
+
+		// using multi-asset cfmm
```
Comment (ValarDragon) [inline x/gamm/pool-models/stableswap/amm_test.go:58]:
```suggestion
		// using multi-asset cfmm (should be equivalent with u = 1, w = 0)
```

File: x/gamm/pool-models/stableswap/amm.go (hunk 3 above)
Comment (ValarDragon) [reply x/gamm/pool-models/stableswap/amm.go:312]: We shouldn't need `u` passed in at all, but thats an optimization we can do in a second pass / shouldn't affect correctness

File: x/gamm/pool-models/stableswap/amm_test.go (hunk 4 above)
Comment (AlpinYukseloglu) [reply x/gamm/pool-models/stableswap/amm_test.go]: Done! Should be in next commit

**Review by ValarDragon — APPROVED** [review]

**Pull request merged**